examples/offline-Q-learning/dqn.py

```diff
@@ -0,0 +1,112 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import warnings
+warnings.simplefilter('default')
+
+import copy
+import paddle.fluid as fluid
+from parl.core.fluid.algorithm import Algorithm
+from parl.core.fluid import layers
+from parl.utils.deprecation import deprecated
+
+__all__ = ['DQN']
+
+
+class DQN(Algorithm):
+    def __init__(self,
+                 model,
+                 hyperparas=None,
+                 act_dim=None,
+                 gamma=None,
+                 lr=None):
+        """ DQN algorithm
+        
+        Args:
+            model (parl.Model): model defining forward network of Q function
+            hyperparas (dict): (deprecated) dict of hyper parameters.
+            act_dim (int): dimension of the action space
+            gamma (float): discounted factor for reward computation.
+            lr (float): learning rate.
+        """
+        self.model = model
+        self.target_model = copy.deepcopy(model)
+
+        if hyperparas is not None:
+            warnings.warn(
+                "the `hyperparas` argument of `__init__` function in `parl.Algorithms.DQN` is deprecated since version 1.2 and will be removed in version 1.3.",
+                DeprecationWarning,
+                stacklevel=2)
+            self.act_dim = hyperparas['action_dim']
+            self.gamma = hyperparas['gamma']
+        else:
+            assert isinstance(act_dim, int)
+            assert isinstance(gamma, float)
+            assert isinstance(lr, float)
+            self.act_dim = act_dim
+            self.gamma = gamma
+            self.lr = lr
+
+    def predict(self, obs):
+        """ use value model self.model to predict the action value
+        """
+        return self.model.value(obs)
+
+    def learn(self, obs, action, reward, next_obs, terminal):
```

Comment:
Most of the codes in `learn` and `supervise_eval` are repeated. Maybe you can define a common function `calc_cost`.